---
title: "Pre-Processing With Phyloseq"
author: "Mark Watson"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  html_document: 
    code_folding: show
    theme: spacelab
    highlight: pygments
    keep_md: no
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
      toc_depth: 3
  keep_md: true  
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.align = "center",
                      fig.path = "BIOMI4300_Project/figures/PreProcessing/") # send any figure output to this folder 
```

# Goals
1. Create S4 object called 'phyloseq'
2. Remove chloroplasts
3. remove mitochondria
4. Remove ASVs from negative controls
5. Evaluate accuracy by ooking at mock community
6. Check for reverse compliments
7. Remove samples without "enough" reads

# Input
1. Metadata
2. ASV table
3. Taxonomy table

# Output
1. s4 phyloseq object

# Setting up the Environment
## Set Seed
```{r set-seed}
set.seed(238428)
```

## Timing Script
```{r timing}
start_time <- Sys.time()
start_time

```

## Load Libraries
```{r load-libraries}
pacman::p_load(tidyverse, BiocManager, devtools, dada2, phyloseq, patchwork, DT, Biostrings, install = FALSE)

```

# Load Data
1. Metadata
```{r load-metadata}

metadata_df <- read_csv("BIOMI4300_Project/data/metadata/SraRunTable.csv", col_names = TRUE)
metadata_df_df <- as.data.frame(metadata_df) %>%
  rename("Run" = "names")

# Subset the columns you want + 'names'
metadata_df_df <- metadata_df_df[, c("names", 
                                    "mussle_community_composition",
                                     "mussle_community_id",
                                     "treatment_id",
                                     "treatment_type")]

# Remove "_org" from treatment_type
metadata_df_df$treatment_type <- gsub("_org", "", metadata_df_df$treatment_type)


# Convert back to sample_data while keeping rownames
rownames(metadata_df_df) <- metadata_df_df$names  # Reset rownames (redundant but safe)
metadata_df <- sample_data(metadata_df_df)



```

2. ASV Table
```{r load-ASV-table}
# First, load asv table, df = dataframe
asv_df <- read.delim(file = "BIOMI4300_Project/data/ASV_counts.tsv", sep = "\t",
                        # Add the column & row names
                        header = TRUE, row.names = 1)

# Fix colnames
colnames(asv_df) <- gsub("_R1_filtered.fastq.gz", "", colnames(asv_df))

# Inspect asv_df
head(asv_df)[,1:5]
```

3. Taxonomy Table
```{r load-tax-table}
# First, load asv table, df = dataframe
tax_df <- read.delim(file = "BIOMI4300_Project/data/ASV_taxonomy.tsv", sep = "\t",
                        # Add the column & row names
                        header = TRUE, row.names = 1)
# Fix colnames
colnames(tax_df) <- gsub("_R1_filtered.fastq.gz", "", colnames(tax_df))
# Inspect asv_df
head(tax_df)[,1:5]

stopifnot(rownames(asv_df) == rownames(tax_df))
```

# Hand off to Phyloseq
```{r phyloseq-handoff}
dim(asv_df)
dim(tax_df)
stopifnot(row.names(asv_df) == row.names(tax_df))

stopifnot(colnames(asv_df) == row.names(metadata_df))
head(colnames(asv_df))
head(row.names(metadata_df))

# Construct the phyloseq object 
raw_physeq <- phyloseq(otu_table(asv_df, taxa_are_rows = TRUE),
                       sample_data(metadata_df),
                       # The taxonomy table needs to first be converted into a matrix
                       tax_table(as.matrix(tax_df)))


# Check the phyloseq object!
raw_physeq

# Save this raw phyloseq object 
save(raw_physeq, file = "BIOMI4300_Project/data/raw_physeq.RData")
```

# Clean up data

There are no controls, no Mitochondria, and no Chloroplasts. Seems that no further cleanup is needed at this point.

# Reverse Compliments

```{r rev-compl}
tax_table <- 
  raw_physeq %>%
  tax_table() %>% 
  data.frame()

# Grab ASV Sequences and Names
asv_names <- tax_table$ASV
asv_seqs <- tax_table$ASVseqs

# Create empty comparison matrix
asv_rev_comp <- matrix(nrow=length(asv_seqs), ncol = length(asv_seqs)) 

# Fix the names in the rows and columns 
rownames(asv_rev_comp) <- asv_names
colnames(asv_rev_comp) <- asv_names

# Convert sequences to Biostrings
asv_dna <- DNAStringSet(asv_seqs) 

# Inspect 
head(asv_dna)

asv_rev_dna <- reverseComplement(asv_dna) 

# Now loop through every asv sequence to check 
for(i in 1:length(asv_seqs)){ # For each asv sequence...
  match_vec <- asv_dna[i] == asv_rev_dna # Compare it to the reverse complement of every other sequence...
  asv_rev_comp[,i] <- match_vec # Write that logical vector as a new column 
}

# Find how many TRUEs (matches) we have, divide by 2 because each pair occurs twice
cat("For", sum(asv_rev_comp) / 2,"ASVs, the reverse complement will need to be removed") 

```
There are no reverse compliments which need to be removed.

# Evaluate Sequencing Depth

```{r sequencing-depth}
# What is the library size/sequencing depth for each sample? 
seqSums_df <- 
  raw_physeq %>%
  otu_table() %>%
  # Sum each sample column 
  colSums() %>%
  data.frame() %>%
  rownames_to_column(var = "names") %>%
  left_join(., metadata_df, by = "names") 

# Rename second column 
colnames(seqSums_df)[2] <- "TotalSeqs"

# check
dim(seqSums_df)

#check depth of samples
seqSums_df %>%
  dplyr::select(names, TotalSeqs) %>%
  arrange(TotalSeqs) %>%
  head()

#plot it!
# plot it as a bar plot 
seqSums_df %>%
  ggplot(aes(x=reorder(names, TotalSeqs), y = TotalSeqs,
             fill = treatment_type)) + 
  geom_bar(stat = "identity") + 
  theme(axis.text.x = element_blank()) + 
  labs(y = "Read Depth", x = "Sample")

# Density plot 
seqSums_df %>%
  ggplot(aes(TotalSeqs, fill = treatment_type)) +
  geom_density(alpha = 0.5) + 
  labs(x = "Read Depth")

```

# Remove samples with few reads
```{r remove-samples}
# view sequencing depth
seqSums_df %>%
  dplyr::select(names, TotalSeqs) %>%
  arrange(TotalSeqs)

# remove low sequenced sample
raw_physeq <- 
  raw_physeq %>%
  subset_samples(., names != "SRR29551974")

```

We removed sample "SRR29551974" for having a very low read depth of 270, when compared to others.

# Save Preprocessed Phyloseq Object
```{r save-physeq}
save(raw_physeq, file = "BIOMI4300_Project/data/processed_physeq.RData")

```

# Final Info for Reproducibility
```{r final-info}
end_time <- Sys.time()
end_time

# elapsed time
elapsed_time <- round((end_time - start_time), 3)
elapsed_time

# Session Info
devtools::session_info()

```