---
title: "Phylogenetic Tree COnstruction"
author: "Mark Watson"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  html_document: 
    code_folding: show
    theme: spacelab
    highlight: pygments
    keep_md: no
    toc: yes
    toc_float:
      collapsed: no
      smooth_scroll: yes
      toc_depth: 3
  keep_md: true  
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.align = "center",
                      fig.path = "BIOMI4300_Project/figures/PhylogeneticTree/") # send any figure output to this folder 
```

# Goals
1. Load in processed phyloseq 
2. create ASV fasta
3. Align Sequences with MAFT
4. Create Tree

# Input
1. proccesed_physeq.RData

# Output
1. ASV fasta
2. ASV alignment
3. Phylogenetic tree

# Setting up the Environment
## Set Seed
```{r set-seed}
set.seed(238428)
```

## Timing Script
```{r timing}
start_time <- Sys.time()
start_time

```

## Load Libraries and Phyloseq Object
```{r load-libraries}
pacman::p_load(phytools, ggtree, RColorBrewer, install = FALSE)

# Load physeq 
load("BIOMI4300_Project/data/processed_physeq.RData")
processed_physeq <- raw_physeq
```

# Write ASV Fasta File
```{r ASV-fasta}
# Pull out ASV sequences and ASV names
asv_seq_df <- 
  processed_physeq@tax_table %>% 
  data.frame() %>% 
  dplyr::select(ASV, ASVseqs)

#View(asv_seq_df)

# Add the ">" to make fasta header
asv_seq_df$ASV <- paste0(">", asv_seq_df$ASV)

#View(asv_seq_df)

# Create fasta object
asv_seq_fasta <- c(rbind(asv_seq_df$ASV, asv_seq_df$ASVseqs))
head(asv_seq_fasta)

# Write to a file 
write(asv_seq_fasta, 
      file = "BIOMI4300_Project/data/preprocessed_ASVs.fasta")
```

# Align the 16S sequences with MAFFT
```{bash MAFFT}
# Write bash code to run MAFT
# First provide the path to MAFT

export PATH=/programs/mafft/bin:$PATH

# change directories to provide the fasta file we made above 

cd data/
pwd

# Set a seed  - using same seed as before for consistency

RANDOM=238428

# Run MAFT
# Windows: control + alt + enter 

/programs/mafft/bin/mafft --auto preprocessed_ASVs.fasta > MAFFT_aligned_ASVs.fasta

# Change back to the project directory 

cd ../../
pwd

```

# FastTreee2
```{bash FastTree2}
# Provide export path to fasttree 
export PATH=/programs/FastTree-2.1.11:$PATH

# cd into the alignment file folder 
cd data/04_Phylogenetic_Tree/
pwd

# Run Fasttree to generate phylogenetic tree 
# parameters: 
    # -nt = indicates it's a nucleotide alignment
    # -gtr = generalized time reversible substitution model 
    # -fastest speed up the model, reduce memory usage (recommended for datasets that have >50,000)
    # -log = output a log file 
    # input alignment file 
    # specify the output tree file 
FastTree -nt -gtr -fastest -log FastTree.log MAFFT_aligned_ASVs.fasta > ASVs_unrooted.tree

# Change back to the main directory 
cd ../../
pwd 
echo "The working directory is $PWD"
```